core/src/main/java/io/grpc/internal/ServiceConfigState.java

```diff
@@ -0,0 +1,130 @@
+
+import io.grpc.Status;
+import javax.annotation.Nullable;
+
+/**
+ * {@link ServiceConfigState} holds the state of the current service config.  It must be mutated
+ * and read from {@link ManagedChannelImpl} constructor or the provided syncContext.
+ */
+final class ServiceConfigState {
+  @Nullable private final ManagedChannelServiceConfig defaultServiceConfig;
+  private final boolean lookUpServiceConfig;
+
+  // mutable state
+  @Nullable private ManagedChannelServiceConfig currentServiceConfig;
+  @Nullable private Status currentError;
+  // Has there been at least one successful update?
+  private boolean updated;
+
+  /**
+   * @param defaultServiceConfig The initial service config, or {@code null} if absent.
+   * @param lookUpServiceConfig {@code true} if service config updates might occur.
+   * @param syncCtx The synchronization context that this is accessed from.
+   */
+  ServiceConfigState(
+      @Nullable ManagedChannelServiceConfig defaultServiceConfig,
+      boolean lookUpServiceConfig) {
+    this.defaultServiceConfig = defaultServiceConfig;
+    this.lookUpServiceConfig = lookUpServiceConfig;
+    if (!lookUpServiceConfig) {
+      this.currentServiceConfig = defaultServiceConfig;
+    }
+  }
+
+  /**
+   * Returns {@code true} if it RPCs should wait on a service config resolution.  This can return
+   * {@code false} if:
+   *
+   * <ul>
+   *   <li>There is a valid service config from the name resolver
+   *   <li>There is a valid default service config and a service config error from the name
+   *       resolver
+   *   <li>No service config from the name resolver, and no intent to lookup a service config.
+   * </ul>
+   *
+   * <p>In the final case, the default service config may be present or absent, and will be the
+   * current service config.
+   */
+  boolean shouldWaitOnServiceConfig() {
+    return !(updated || !expectUpdates());
+  }
+
+  /**
+   * Gets the current service config.
+   *
+   * @throws IllegalStateException if the service config has not yet been updated.
+   */
+  @Nullable ManagedChannelServiceConfig getCurrentServiceConfig() {
+    checkState(!shouldWaitOnServiceConfig(), "still waiting on service config");
+    return currentServiceConfig;
+  }
+
+  /**
+   * Gets the current service config error.
+   *
+   * @throws IllegalStateException if the service config has not yet been updated.
+   */
+  @Nullable Status getCurrentError() {
+    checkState(!shouldWaitOnServiceConfig(), "still waiting on service config");
+    return currentError;
+  }
+
+  /**
+   * Returns {@code true} if the update was successfully applied, else {@code false}.
+   */
+  boolean update(Status error) {
+    checkNotNull(error, "error");
+    checkArgument(!error.isOk(), "can't use OK error");
+    checkState(expectUpdates(), "unexpected service config update");
+    boolean firstUpdate = !updated;
+    updated = true;
+    if ((firstUpdate && defaultServiceConfig == null) || currentError != null) {
```

Comment:
The flow of control is a little difficult to follow. It would be more readable if you structure it around the input conditions rather than output values:
```java
if (firstUpdate) {
  if (defaultServieConfig != null) {
    currentCoe = ConfigOrError.fromConfig(defaultServiceConfig);
  } else {
    currentCoe = ConfigOrError.fromError(error);
  }
} else {
  // Subsequent updates. Update the error only if already in error
  // Discard it otherwise.
  if (currentCoe != null && currentCoe.error != null) {
    currentCoe = ConfigOrError.fromError(error);
  }
}
```

Reply:
I tried it out as the folllowing, but I think its much more complicated.  It's not defined in terms of inputs or outputs, but as the state diagram.  You suggestion is in terms of edges, while this is in terms of the nodes.  The first condition (as is) says if it is in state 1 or state 2, go to state 2.  

```java
    boolean firstUpdate = !updated;
    updated = true;
    if (firstUpdate) {
      if (defaultServiceConfig == null) {
        assert currentServiceConfig == null;
        currentError = error;
        return true;
      } else if (currentServiceConfig == null) {
        currentServiceConfig = defaultServiceConfig;

      }
      return false;
    } else {
      if (currentError != null) {
        assert currentServiceConfig == null;
        currentError = error;
        return true;
      } else {
        return false;
      }
    }
```

Reply:
I find edge-oriented is easier to reason about, because you can easily put comments on the edges. Node-oriented is harder to read, unless you have the state diagram at hand, which is not the case here.

Reply:
I updated the code as requested, but I think it is much more verbose, more complicated, and not as obviously correct.  I left the old functions in place to work with the unit tests.